internal/typename/gotypes/typesname.go

```diff
@@ -0,0 +1,142 @@
+package gotypes
+
+import (
+	"fmt"
+	"go/types"
+)
+
+// NameOf returns the fully-qualified name of the given type.
+func NameOf(t types.Type) string {
+	switch t := t.(type) {
+	case *types.Basic:
+		return t.String()
+	case *types.Pointer:
+		return t.String()
+	case *types.Slice:
+		return t.String()
+	case *types.Array:
+		return t.String()
+	case *types.Map:
+		return t.String()
+	case *types.Chan:
+		return buildChanName(t)
+	case *types.Interface:
+		return buildInterfaceName(t)
+	case *types.Named:
+		return t.String()
+	case *types.Struct:
+		return buildStructName(t)
+	case *types.Signature:
+		return "func" + buildFuncSignature(t)
+	}
+	// COVERAGE NOTE: this panic is only possible if additional implementation
+	// of go/types.Type interface is introduced to represent the first-class type
+	// in Go. Not all go/types.Type interface implementations represent
+	// first-class types. For instance, go/types.Tuple can only be a part of
+	// signatures or multiple assignments.
+	panic(fmt.Sprintf("unknown type %s", t))
```

Comment:
```suggestion

	// COVERAGE: This panic is only possible if an additional implementation of
	// the types.Type interface is introduced to represent a new first-class
	// type in Go. Not all implementations of types.Type represent first-class
	// types. For instance, types.Tuple can only be a part of signatures or
	// multiple assignments.
	panic(fmt.Sprintf("unknown type %s", t))
```

This looks like I changed everything because I reflowed the line wrapping, but basically I:

- Fixed a few articles
- Refer to interface as `types.Type`
- Removed some duplicate words